server/storage/endpoint/keyspace.go

```diff
@@ -0,0 +1,104 @@
+// Copyright 2022 TiKV Project Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package endpoint
+
+import (
+	"strconv"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/pingcap/kvproto/pkg/keyspacepb"
+	"go.etcd.io/etcd/clientv3"
+)
+
+const spaceIDBase = 10
+
+type KeyspaceStorage interface {
+	// SaveKeyspace saves the given keyspace to the storage.
+	SaveKeyspace(*keyspacepb.KeyspaceMeta) error
+	// LoadKeyspace loads keyspace specified by spaceID.
+	LoadKeyspace(spaceID uint32, keyspace *keyspacepb.KeyspaceMeta) (bool, error)
+	// RemoveKeyspace removes target keyspace specified by spaceID.
+	RemoveKeyspace(spaceID uint32) error
+	// LoadRangeKeyspace loads no more than limit keyspaces starting at startID.
+	LoadRangeKeyspace(startID uint32, limit int) ([]*keyspacepb.KeyspaceMeta, error)
+	// SaveKeyspaceID saves keyspace name to ID lookup information.
+	SaveKeyspaceID(spaceID uint32, name string) error
+	// LoadKeyspaceID loads keyspace ID for the given keyspace specified by name.
+	LoadKeyspaceID(name string) (bool, uint32, error)
+}
+
+var _ KeyspaceStorage = (*StorageEndpoint)(nil)
+
+// SaveKeyspace saves the given keyspace to the storage.
+func (se *StorageEndpoint) SaveKeyspace(keyspace *keyspacepb.KeyspaceMeta) error {
```

Comment:
We should refuse SaveKeyspace when the state is ```ARCHIVED```

Reply:
input validation and concurrency control are in KeyspaceManager (tikv/pd#5291), KeyspaceStorage here simply provides a way to interact with underlying storageEndpoint.